src/scanner.c

```diff
@@ -1,127 +1,170 @@
+	char * tmp = NULL; /* shut up, GCC */
+	size_t imgLen = 0;
+
+	/*
+	 * If you hate me now, I understand.
+	 * just wait till the parser...
+	 */
+
+
+	/*
+	 * optimum strategy, filter by ascii ranges, separate single char
+	 * tokens from multi char tokens to avoid redudant checks?
+	 *     ~Eric B.
+	 */
+	if (matches_obracket(*source))
+		__set_char_tok(source, OBRACKET);
+	else if (matches_cbracket(*source))
+		__set_char_tok(source, CBRACKET);
+	else if (matches_obrace(*source))
+		__set_char_tok(source, OBRACE);
+	else if (matches_cbrace(*source))
+		__set_char_tok(source, CBRACE);
+	else if (matches_oparen(*source))
+		__set_char_tok(source, OPAREN);
+	else if (matches_cparen(*source))
+		__set_char_tok(source, CPAREN);
+	else if (matches_plus(*source))
+		__set_char_tok(source, PLUS);
+	else if (matches_minus(*source))
+		__set_char_tok(source, MINUS);
+	else if (matches_star(*source))
+		__set_char_tok(source, STAR);
+	else if (matches_semicolon(*source))
+		__set_char_tok(source, SEMICOLON);
+	else if (matches_underscore(*source))
+		__set_char_tok(source, UNDERSCORE);
+	else if (matches_bang(*source))
+		__set_char_tok(source, BANG);
+	else if (matches_func(*source))
+		__set_nlen_tok(source, 4, FUNC);
+	else if (matches_comma(*source))
+		__set_char_tok(source, COMMA);
+	else if (matches_fslash(*source))
+		__set_char_tok(source, FSLASH);
+	else if (matches_bslash(*source))
+		__set_char_tok(source, BSLASH);
+	else if (matches_ampersand(*source))
+		__set_char_tok(source, AMPERSAND);
+	else if (matches_pipe(*source))
+		__set_char_tok(source, PIPE);
+	else if (matches_at(*source))
+		__set_char_tok(source, AT);
+	else if (matches_quote(*source))
+		__set_char_tok(source, QUOTE);
+	else if (matches_apostrophe(*source))
+		__set_char_tok(source, APOSTROPHE);
+	else if (matches_capture(*source))
+		__set_nlen_tok(source, 7, CAPTURE);
+	else if (matches_maybe(*source))
+		__set_nlen_tok(source, 5, MAYBE);
+	else if (matches_int(*source))
+		__set_nlen_tok(source, 3, INT);
+	else if (matches_bool(*source))
+		__set_nlen_tok(source, 4, BOOL);
+	else if (matches_char(*source))
+		__set_nlen_tok(source, 4, CHAR);
+	else if (matches_double(*source))
+		__set_nlen_tok(source, 6, DOUBLE);
+	else if (matches_long(*source))
+		__set_nlen_tok(source, 4, LONG);
+	else if (matches_unsigned(*source))
+		__set_nlen_tok(source, 8, UNSIGNED);
+	else if (matches_question(*source))
+		__set_char_tok(source, QUESTION);
+	else if (matches_and(*source))
+		__set_nlen_tok(source, 3, AND);
+	else if (matches_or(*source))
+		__set_nlen_tok(source, 2, OR);
+	else if (matches_just(*source))
+		__set_nlen_tok(source, 4, JUST);
+	else if (matches_notoken(*source) || (**source) == '\0')
+		__set_char_tok(source, NOTOKEN);
+	else if ((imgLen = matches_integer(*source, &tmp))) {
+		token->image = tmp;
+		token->tok = INTEGER;
+		*source += imgLen;
+	} else if ((imgLen = matches_float(*source, &tmp)))
+		__set_nlen_tok(source, imgLen, FLOAT);
+	else {
```

Comment:
Doesn't the parser need a general identifier token, basically an unquoted string, for variable names? I could be missing that, but I don't see it here.
There is a `case ALPHA` on `print_each_node`, maybe it should be using that?
Is it a special case in whatever calls this function, so it is ignored here?

Reply:
We haven't implemented the identifier token yet. That's the next TODO.